### ISSUE TITLE
Bulk Export

### ISSUE THREAD

**Issue by RJGhugo** [darktable-org/darktable#10370]
**Bulk Export**
Exporting multiple images in one go.
**no multithreading**
darktable is only using one thread. The job is done sequential not paralell.

**Comment by MStraeten** [comment]
The processing while export of each file is already highly parallelized - you can see this if you have a look at the cpu usage per core …

**Comment by Nilvus** [comment]
Be sure too to set correctly cpu/gpu/memory settings in processing tab in main preferences.

**Comment by RJGhugo** [comment]
**System Infos**
uname -a
Linux workstation 5.14.16-arch1-1 #1 SMP PREEMPT Tue, 02 Nov 2021 22:22:59 +0000 x86_64 GNU/Linux

darktable --version
this is darktable 3.6.1
copyright (c) 2009-2021 johannes hanika
darktable-dev@lists.darktable.org

compile options:
  bit depth is 64 bit
  normal build
  SSE2 optimized codepath enabled
  OpenMP support enabled
  OpenCL support enabled
  Lua support enabled, API version 7.0.0
  Colord support enabled
  gPhoto2 support enabled
  GraphicsMagick support enabled
  ImageMagick support disabled
  OpenEXR support enabled

**CPU/GPU/Memory**
![darktable-CPU-GPU-Speicher](https://user-images.githubusercontent.com/41795683/141079900-02ae9bd4-b391-4f8e-a307-b245196ba445.png)
![darktable-CPU-usage-2](https://user-images.githubusercontent.com/41795683/141079942-d1166217-0764-4f09-8c01-931b0874670a.png)
![gpu-short peak](https://user-images.githubusercontent.com/41795683/141081124-b4bc3e01-4a04-4cd6-8b52-98a5e43a2376.png)
**Performance Tuning**
https://docs.darktable.org/usermanual/3.6/special-topics/opencl/performance/

nano ~/.config/darktable/darktablerc
opencl_async_pixelpipe=false
opencl_number_event_handles=25 --> 100
opencl_synch_cache=active module
opencl_micro_nap=1000
opencl_use_pinned_memory=false --> tru
e

**Comment by RJGhugo** [comment]
Export: RAW --> TIFF (16bit)

**Comment by wpferguson** [comment]
I think the export process is probably GPU bound when using opencl.  Therefore, 1 CPU is feeding data to the GPU, then getting it back and writing to the file.  To utilize more threads you would need more GPUs.  

I wonder if switching off opencl for export would result in an overall speed up?  While the GPU can process 1 image faster than the CPU,  can multiple CPUs (threads) process multiple images quicker than running every image through the GPU?

**Comment by RJGhugo** [comment]
Thanks for the hint. I do not have a significant less exporting time while exporting RAW to TIFF (16bit).

![cpu while export TIFF mit Pfeil](https://user-images.githubusercontent.com/41795683/141183443-04ad8f44-18e9-4e75-9415-b57d5d1d3a59.png)
![no OpenCL mit Pfeil](https://user-images.githubusercontent.com/41795683/141183446-80471ab5-170b-411c-90bc-5a475707b0bf.png)

**Comment by MStraeten** [comment]
maybe switch your opencl setting to „Standard“ to be able to tweak your opencl settings. (https://docs.darktable.org/usermanual/3.6/special-topics/opencl/scheduling-profile/)

**Comment by RJGhugo** [comment]
Thanks. I tried to run with and without openCL. The idle time is still there between each export. Each export may be highly parallelized (see peak at starting of exporting a picture), but the **bulk export** it self seems to be a **sequential task**...

with openCL:
![darktable-CPU-usage-mit-openCL mit Pfeil](https://user-images.githubusercontent.com/41795683/141855152-ccce5a9f-e532-4272-b420-06c9d5316353.png)
without openCL:
![darktable-CPU-usage-ohne-openCL mit Pfeil](https://user-images.githubusercontent.com/41795683/141855157-d627e5b4-8d37-41f2-844d-43fc81d7bc98.png)
..

**Comment by wpferguson** [comment]
Earlier I said that it was GPU limited, but I should have said pipeline limited.  You can only run 1 copy of the pipeline at a time.

**Comment by RJGhugo** [comment]
Thanks. This means, that there is no way to speed up export by parallelizing?

Shoud I close the issue?

**Comment by wpferguson** [comment]
There is another issue (or discussion) about displaying 2 processed images side by side and zooming and panning.  The problem with implementing that is you need 2 pipelines running in parallel.

Until someone figures out how to run more than one pipeline at a time (and/or there's enough resources to make it possible), this isn't going to happen.  But enough people are raising issues that it's at least on the radar.

**Comment by aurelienpierre** [comment]
No benefit will appear from running export jobs in parallel, the image processing itself is already parallel. Unless you have more than one GPU, which is not common.

**Comment by RJGhugo** [comment]
Thanks.
I can life with it.
I was only wondering about the short CPU/GPU peak comparing with the idle time.

**Comment by PeterWem** [comment]
Compression?
![Skärmbild från 2021-12-24 14-01-38](https://user-images.githubusercontent.com/34200649/147354612-e2bd219d-211b-4577-8ad1-f9542aeb553c.png)
![Skärmbild från 2021-12-24 14-00-35](https://user-images.githubusercontent.com/34200649/147354615-e362b0b4-2344-422c-bc5c-7b2a31dd2265.png)
![Skärmbild från 2021-12-24 13-58-43](https://user-images.githubusercontent.com/34200649/147354617-a506940e-fa57-4350-ac80-a69b6e416656.png)
![Skärmbild från 2021-12-24 13-57-33](https://user-images.githubusercontent.com/34200649/147354619-936a6445-b203-4eb8-b493-e21e700da2ab.png)

**Comment by kmilos** [comment]
Compression could indeed be apart of it. For TIFF export, [you probably want a recent libtiff built against libdeflate](https://aras-p.info/blog/2021/08/09/EXR-libdeflate-is-great/).

**Comment by GustavHaapalahti** [comment]
When I do bulk exports my CPU usage is at 3% and GPU usage is at 0-1%
I think multiple export threads would definitely speed up processing at least for my computer.
Geforce 5070Ti and AMD 9950X3D CPU.

**Comment by TurboGit** [comment]
Probably not as it will soon eat all memory.
